File: feature/permission/src/iosMain/swift/presentation/location/LocationPermissionScreen.swift
import SwiftUI
import KMPObservableViewModelSwiftUI
import iosBridge

struct LocationPermissionScreen: View {
    let onPermissionGrant: () -> Void

    @StateViewModel private var viewModel = provideLocationPermissionViewModel()
    @State private var authorization = LocationAuthorization()

    var body: some View {
        // SKIE turns `states` into an AsyncSequence, so SwiftUI can observe the store directly.
        Observing(viewModel.states) { state in
            LocationPermissionContent(
                state: state,
                onAllow: { viewModel.onIntent(intent: LocationPermissionIntentAllow.shared) },
                onOpenSettings: { viewModel.onIntent(intent: LocationPermissionIntentOpenSettings.shared) }
            )
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        // `.task` runs while the screen is visible, so the store sees the subscriber come and go.
        .task {
            for await action in viewModel.actions {
                switch onEnum(of: action) {
                case .launchPermissionRequest:
                    authorization.request()
                case .launchSettings:
                    if let url = URL(string: UIApplication.openSettingsURLString) {
                        _ = await UIApplication.shared.open(url)
                    }
                case .permissionGranted:
                    onPermissionGrant()
                }
            }
        }
        // Reports the first status, and every change, e.g. after the user allows it in Settings.
        .task {
            for await status in authorization.statuses {
                viewModel.onIntent(intent: LocationPermissionIntentStatusChanged(status: status))
            }
        }
    }
}

private struct LocationPermissionContent: View {
    let state: LocationPermissionState
    let onAllow: () -> Void
    let onOpenSettings: () -> Void

    var body: some View {
        // `onEnum(of:)` makes the sealed interface exhaustive – a new state stops compiling here.
        switch onEnum(of: state) {
        case .checking:
            Color.clear
        case .notDetermined:
            PermissionMessage(
                message: "KMP Showcase uses your approximate location to show the weather where you are.",
                buttonTitle: "Allow location access",
                action: onAllow
            )
        case .denied:
            PermissionMessage(
                message: "Without location access KMP Showcase can't show the weather where you are.",
                buttonTitle: "Try again",
                action: onAllow
            )
        case .permanentlyDenied:
            PermissionMessage(
                message: "Location access is turned off for KMP Showcase. Allow it in Settings to continue.",
                buttonTitle: "Open Settings",
                action: onOpenSettings
            )
        case .servicesDisabled:
            PermissionMessage(
                message: "Location Services are turned off. Turn them on in Settings › Privacy & Security › Location Services.",
                buttonTitle: "Open Settings",
                action: onOpenSettings
            )
        case .restricted:
            PermissionMessage(
                message: "Location access is restricted on this device, e.g. by parental controls or device management."
            )
        }
    }
}

/// An explanation plus an optional button; the button is hidden when the user can't act.
private struct PermissionMessage: View {
    let message: String
    var buttonTitle: String?
    var action: () -> Void = {}

    var body: some View {
        VStack(spacing: 8) {
            Text("📍")
                .font(.system(size: 40))
            Text(message)
                .multilineTextAlignment(.center)
            if let buttonTitle {
                Button(buttonTitle, action: action)
                    .buttonStyle(.borderedProminent)
            }
        }
        .padding(24)
    }
}

#Preview("Not determined") {
    LocationPermissionContent(state: LocationPermissionStateNotDetermined.shared, onAllow: {}, onOpenSettings: {})
}

#Preview("Denied") {
    LocationPermissionContent(state: LocationPermissionStateDenied.shared, onAllow: {}, onOpenSettings: {})
}

#Preview("Permanently denied") {
    LocationPermissionContent(state: LocationPermissionStatePermanentlyDenied.shared, onAllow: {}, onOpenSettings: {})
}

#Preview("Services disabled") {
    LocationPermissionContent(state: LocationPermissionStateServicesDisabled.shared, onAllow: {}, onOpenSettings: {})
}

#Preview("Restricted") {
    LocationPermissionContent(state: LocationPermissionStateRestricted.shared, onAllow: {}, onOpenSettings: {})
}

File: feature/permission/src/androidMain/kotlin/com/igorwojda/showcase/feature/permission/presentation/location/LocationPermissionScreen.kt
package com.igorwojda.showcase.feature.permission.presentation.location

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.tooling.preview.PreviewParameter
import androidx.compose.ui.tooling.preview.PreviewParameterProvider
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import org.koin.androidx.compose.koinViewModel
import pro.respawn.flowmvi.compose.dsl.subscribe

@Composable
fun LocationPermissionScreen(
    onPermissionGranted: () -> Unit,
    viewModel: LocationPermissionViewModel = koinViewModel(),
) {
    val store = viewModel.store

    // Reports the first status, and every change, e.g. after the user allows it in Settings.
    val requester = rememberLocationPermissionRequester { status ->
        store.intent(LocationPermissionIntent.StatusChanged(status))
    }

    // The lambda consumes MVIActions as they arrive; it only runs while the UI is visible.
    val state by store.subscribe { action ->
        when (action) {
            LocationPermissionAction.LaunchPermissionRequest -> requester.request()
            LocationPermissionAction.LaunchSettings -> requester.openSettings()
            LocationPermissionAction.PermissionGranted -> onPermissionGranted()
        }
    }

    Scaffold { contentPadding ->
        Box(
            modifier = Modifier
                .fillMaxSize()
                .padding(contentPadding),
        ) {
            LocationPermissionContent(
                state = state,
                onAllowClick = { store.intent(LocationPermissionIntent.Allow) },
                onOpenSettingsClick = { store.intent(LocationPermissionIntent.OpenSettings) },
                modifier = Modifier.align(Alignment.Center),
            )
        }
    }
}

@Composable
private fun LocationPermissionContent(
    state: LocationPermissionState,
    onAllowClick: () -> Unit,
    onOpenSettingsClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    when (state) {
        LocationPermissionState.Checking -> Unit

        LocationPermissionState.NotDetermined -> PermissionMessage(
            message = "KMP Showcase uses your approximate location to show the weather where you are.",
            buttonText = "Allow location access",
            onButtonClick = onAllowClick,
            modifier = modifier,
        )

        LocationPermissionState.Denied -> PermissionMessage(
            message = "Without location access KMP Showcase can't show the weather where you are.",
            buttonText = "Try again",
            onButtonClick = onAllowClick,
            modifier = modifier,
        )

        LocationPermissionState.PermanentlyDenied -> PermissionMessage(
            message = "Location access is turned off for KMP Showcase. Allow it in Settings to continue.",
            buttonText = "Open Settings",
            onButtonClick = onOpenSettingsClick,
            modifier = modifier,
        )

        LocationPermissionState.ServicesDisabled -> PermissionMessage(
            message = "Location is turned off on this device. Turn it on in Settings to continue.",
            buttonText = "Open Settings",
            onButtonClick = onOpenSettingsClick,
            modifier = modifier,
        )

        LocationPermissionState.Restricted -> PermissionMessage(
            message = "Location access is restricted on this device, e.g. by parental controls or device management.",
            buttonText = null,
            onButtonClick = {},
            modifier = modifier,
        )
    }
}

/** An explanation plus an optional [buttonText] button; the button is hidden when the user can't act. */
@Composable
private fun PermissionMessage(
    message: String,
    buttonText: String?,
    onButtonClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier = modifier.padding(all = 24.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Text(text = "📍", fontSize = 40.sp)
        Text(text = message, textAlign = TextAlign.Center)
        if (buttonText != null) {
            Button(onClick = onButtonClick) {
                Text(buttonText)
            }
        }
    }
}

private class LocationPermissionStateProvider : PreviewParameterProvider<LocationPermissionState> {
    override val values = sequenceOf(
        LocationPermissionState.NotDetermined,
        LocationPermissionState.Denied,
        LocationPermissionState.PermanentlyDenied,
        LocationPermissionState.ServicesDisabled,
        LocationPermissionState.Restricted,
    )
}

@Preview(showBackground = true)
@Composable
private fun LocationPermissionContentPreview(
    @PreviewParameter(LocationPermissionStateProvider::class) state: LocationPermissionState,
) {
    MaterialTheme {
        LocationPermissionContent(state = state, onAllowClick = {}, onOpenSettingsClick = {})
    }
}
